crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1416,10 +1416,31 @@ impl<'db> ClassType<'db> {
         let DefinitionKind::Class(class_stmt_node) = definition.kind(db) else {
             panic!("Class type definition must have DefinitionKind::Class");
         };
-        class_stmt_node
-            .bases()
-            .iter()
-            .map(move |base_expr| definition_expression_ty(db, definition, base_expr))
+
+        let has_type_params = class_stmt_node.type_params.is_some();
+
+        let type_scope_info: Option<(ScopeId<'db>, &TypeInference<'db>)> = if has_type_params {
+            let file = definition.file(db);
+            let index = semantic_index(db, file);
+            // we need to use the type param'd scope
+            let type_param_scope = index
+                .node_scope(NodeWithScopeRef::ClassTypeParameters(class_stmt_node))
+                .to_scope_id(db, file);
```

Comment:
It's unclear to me what the cost of the `semantic_index` lookup would be in practice. I just want to know what scope I need to be looking at. While `ClassType` does have `body_scope`, that is a separate scope from the type parameter scope.

Would it make sense to add `bases_scope` to `ClassType` here to avoid needing this index build up?

Reply:
I would expect that we could reuse the `self.file` and `self.index` because all the expressions in the class's base should be in the same file and from the same index. 

However, we probably don't want to call `infer_scope_types` because inferring the class's bases then suddenly becomes dependent of **all** types in that other scope, resulting in poor incremental caching. 

I'm not quiet sure but maybe `definition_expression_ty` is a better fit?

Reply:
Class bases aren't a definition, so we can't query on a definition type.

We could mark the class bases as a "standalone expression" (or set of them) so we can query for their types directly, but I don't think this is worth the extra Salsa tracked structs. The "type params scope" is automatically generated and implicit and contains nothing but definitions for type parameters, and the class bases/keywords. There isn't "all types in that scope" to be worried about, because there are no other types in that scope. So I think `infer_scope_types` is the best option here.

Reply:
> I would expect that we could reuse the self.file and self.index

We aren't in `TypeInferenceBuilder` here, so those aren't available.

Reply:
I think perhaps the simpler way to get the type params scope is that it must always be the parent scope of the body scope (for a class with type params.) We should be able to add a `ScopeId::parent` method (it would need to use the symbol table, but not the semantic index.)

I'm also not opposed to adding `bases_scope` (or optional `type_params_scope`) to `ClassType`.

Reply:
You can also use `index.scopes_by_node` when the query already depends on the index anyway (which `parent` would as well?)

Reply:
`ScopeId::parent` would rely on just the symbol table, which is part of the index, but in principle could be backdated independently if it didn't change? Unlikely though; I don't think depending on the index here is a problem.

Reply:
Ended up storing the bases scope in `ClassType` (which gets built up in the inference builder so we have access to the index right there). Definitely feels right after having written it up.